File: plus-f/src/main/scala/com/socialthingy/p2p/Peer.scala
package com.socialthingy.p2p

import java.net._
import java.nio.{ByteBuffer, ByteOrder}
import java.util.concurrent.{Executors, TimeUnit}

import com.codahale.metrics.{Histogram, Meter, ExponentiallyDecayingReservoir}
import org.slf4j.LoggerFactory

import scala.concurrent.duration.FiniteDuration
import scala.util.control.NonFatal
import scala.util.{Failure, Success, Try}

object Peer {
  sealed trait State
  case object Initial extends State
  case object WaitingForPeer extends State
  case object Connected extends State
  case object Closing extends State
}

class Peer(bindAddress: InetSocketAddress,
           discoveryServiceAddress: InetSocketAddress,
           callbacks: Callbacks,
           serialiser: Serialiser,
           deserialiser: Deserialiser,
           timeout: FiniteDuration) {
  import PacketUtils._
  import Peer._

  type Handler = (ByteBuffer, InetSocketAddress) => Unit

  val log = LoggerFactory.getLogger(classOf[Peer])
  var socket: DatagramSocket = _
  val socketHandlerExecutor = Executors.newSingleThreadExecutor()
  implicit val byteOrder = ByteOrder.BIG_ENDIAN

  val latencies = new Histogram(new ExponentiallyDecayingReservoir)
  val sizes = new Histogram(new ExponentiallyDecayingReservoir)
  val meter = new Meter()
  var outOfOrder = 0

  var peerConnection: Option[InetSocketAddress] = None

  val states = Map[State, Handler](
    Initial -> doNothing _,
    WaitingForPeer -> waitForResponse _,
    Connected -> connectedToPeer _
  )

  var currentState: State = Initial
  var currentSessionId: Option[String] = None
  var lastReceivedTimestamp = -1L

  def doNothing(data: ByteBuffer, source: InetSocketAddress): Unit = ()

  def connectedToPeer(data: ByteBuffer, source: InetSocketAddress): Unit = {
    Try {
      WrappedData(decompress(data), deserialiser)
    } match {
      case Success(decompressed) =>
        latencies.update(System.currentTimeMillis - decompressed.systemTime)
        meter.mark()
        if (decompressed.timestamp >= lastReceivedTimestamp) {
          lastReceivedTimestamp = decompressed.timestamp
          callbacks.data(decompressed.content)
        } else {
          outOfOrder = outOfOrder + 1
        }

        peerConnection foreach { x =>
          if (!x.equals(source))
            log.info(s"Switching peer connection from ${x.toString} to ${source.toString}")
        }

        peerConnection = Some(source)

      case Failure(ex) =>
        log.error(
          s"Unable to decode received message ${data.toString} from ${source.toString}",
          ex
        )
    }
  }

  def waitForResponse(data: ByteBuffer, source: InetSocketAddress): Unit = {
    val result = new String(data.array(), data.position(), data.remaining(), "UTF-8")
    result.split('|').toList match {
      case "PEER" :: peerHost :: peerPort :: Nil =>
        callbacks.connectedToPeer(peerPort.toInt)
        log.info(s"Connected to peer at $peerHost:$peerPort")
        currentState = Connected
        peerConnection = Some(new InetSocketAddress(peerHost, peerPort.toInt))

      case "WAIT" :: Nil =>
        callbacks.waitingForPeer()
        log.info("Waiting for peer to join")

      case _ =>
        log.error(
          s"Unrecognised message received from ${source.toString} with content [${data.toString}], still waiting for peer"
        )
    }

  }

  def startIfRequired(): Unit = {
    if (socket == null || socket.isClosed) {
      socket = new DatagramSocket(bindAddress)
      socket.setSoTimeout(timeout.toMillis.toInt)

      socketHandlerExecutor.submit(new Runnable {
        def run() = {
          val receivedPacket = new DatagramPacket(Array.ofDim[Byte](16384), 16384)
          while (!socket.isClosed) {
            try {
              socket.receive(receivedPacket)
              log.debug("Received packet of size {} from {}", receivedPacket.getLength, receivedPacket.getAddress.toString)
              sizes.update(receivedPacket.getLength)
              states(currentState)(
                ByteBuffer.wrap(receivedPacket.getData, receivedPacket.getOffset, receivedPacket.getLength),
                receivedPacket.getSocketAddress.asInstanceOf[InetSocketAddress]
              )
            } catch {
              case e: SocketTimeoutException if currentState == WaitingForPeer =>
                callbacks.discoveryTimeout()
                close()

              case e: SocketException if e.getMessage == "Socket closed" => // do nothing, expected
              case NonFatal(e) => log.error("Failure in receive loop", e)
            }
          }
          log.info("Socket closed")
        }
      })
    }
  }

  def join(sessionId: String, fwdPort: Option[Int] = None): Unit = {
    startIfRequired()

    callbacks.discovering()
    val joinCommand = fwdPort match {
      case Some(p) => s"JOIN|$sessionId|$p"
      case None => s"JOIN|$sessionId"
    }

    currentState = WaitingForPeer
    currentSessionId = Some(sessionId)
    socket.send(buildPacket(joinCommand, discoveryServiceAddress))
  }

  def send(data: RawData): Unit = (peerConnection, socket) match {
    case (None, _) => log.error("Unable to send data, no peer connection")
    case (_, s) if s.isClosed => log.error("Unable to send data, socket closed")
    case (Some(conn), s) =>
      val compressed = compress(data.wrap.pack(serialiser))
      log.debug("Sending packet of size {} to {}", compressed.remaining(), conn.toString)
      s.send(new DatagramPacket(compressed.array(), compressed.position(), compressed.remaining(), conn))
  }

  def statistics(): Statistics = Statistics(
    latencies.getSnapshot.get99thPercentile.toInt,
    outOfOrder,
    sizes.getSnapshot.get99thPercentile.toInt,
    meter.getOneMinuteRate
  )

  def close(): Unit = (currentState, currentSessionId) match {
    case (WaitingForPeer, Some(sessionId)) =>
      socket.send(buildPacket(s"CANCEL|$sessionId", discoveryServiceAddress))
      reset()
      callbacks.discoveryCancelled()

    case _ =>
      reset()
      callbacks.closed()
  }

  private def reset() = {
    if (socket != null) {
      socket.close()
    }
    currentSessionId = None
    currentState = Initial
    peerConnection = None
  }
}

case class Statistics(latency: Int, outOfOrder: Int, size: Int, oneMinuteRate: Double)

File: plus-f/src/main/kotlin/com/socialthingy/p2p/Peer.kt
package com.socialthingy.p2p

import java.net.*
import java.nio.ByteBuffer
import java.util.concurrent.Executors

import com.codahale.metrics.ExponentiallyDecayingReservoir
import com.codahale.metrics.Histogram
import com.codahale.metrics.Meter
import org.slf4j.LoggerFactory
import java.util.*

enum class State {
    INITIAL, WAITING_FOR_PEER, CONNECTED, CLOSING
}

class Peer(private val bindAddress: InetSocketAddress,
           private val discoveryServiceAddress: InetSocketAddress,
           val callbacks: Callbacks,
           val serialiser: Serialiser,
           val deserialiser: Deserialiser,
           private val timeout: java.time.Duration) {

    val log = LoggerFactory.getLogger(javaClass)!!
    private var socket: Optional<DatagramSocket> = Optional.empty()
    private val socketHandlerExecutor = Executors.newSingleThreadExecutor()!!

    private val latencies = Histogram(ExponentiallyDecayingReservoir())
    private val sizes = Histogram(ExponentiallyDecayingReservoir())
    private val meter = Meter()
    private var outOfOrder = 0

    private var peerConnection: Optional<InetSocketAddress> = Optional.empty()

    private val states = HashMap<State, (ByteBuffer, InetSocketAddress) -> Unit>()

    init {
        states.put(State.INITIAL, this::doNothing)
        states.put(State.WAITING_FOR_PEER, this::waitForResponse)
        states.put(State.CONNECTED, this::connectedToPeer)
    }

    var currentState: State = State.INITIAL
    var currentSessionId: Optional<String> = Optional.empty()
    var lastReceivedTimestamp = -1L

    private fun doNothing(data: ByteBuffer, source: InetSocketAddress) {}

    private fun connectedToPeer(data: ByteBuffer, source: InetSocketAddress) {
        try {
            val decompressed = WrappedData(PacketUtils.decompress(data), deserialiser)
            latencies.update(System.currentTimeMillis() - decompressed.systemTime)
            meter.mark()
            if (decompressed.timestamp >= lastReceivedTimestamp) {
                lastReceivedTimestamp = decompressed.timestamp
                callbacks.data(decompressed.content)
            } else {
                outOfOrder += 1
            }

            peerConnection.ifPresent { x ->
                if (x != source) {
                    log.info("Switching peer connection from {} to {}", x.toString(), source.toString())
                }
            }

            peerConnection = Optional.of(source)

        } catch (ex: Exception) {
            log.error(
                    String.format("Unable to decode received message %s from %s", data.toString(), source.toString()),
                    ex
            )
        }
    }

    private fun waitForResponse(data: ByteBuffer, source: InetSocketAddress) {
        val result = String(data.array(), data.position(), data.remaining())
        val splitMessage = result.split('|')

        when {
            splitMessage.size == 3 && splitMessage[0] == "PEER" -> {
                val peerHost = splitMessage[1]
                val peerPort = splitMessage[2].toInt()
                callbacks.connectedToPeer(peerPort)
                log.info("Connected to peer at {}:{}", peerHost, peerPort)
                currentState = State.CONNECTED
                peerConnection = Optional.of(InetSocketAddress(peerHost, peerPort))
            }

            splitMessage.size == 1 && splitMessage[0] == "WAIT" -> {
                callbacks.waitingForPeer()
                log.info("Waiting for peer to join")
            }

            else -> {
                log.error(
                        "Unrecognised message received from {} with content {}, still waiting for peer",
                        source.toString(),
                        data.toString()
                )
            }
        }
    }

    private fun startIfRequired() {
        if (socket.map { it.isClosed }.orElse(false)) {
            val newSocket = DatagramSocket(bindAddress)
            socket = Optional.of(newSocket)
            newSocket.soTimeout = timeout.toMillis().toInt()

            socketHandlerExecutor.submit({
                val receivedPacket = DatagramPacket(ByteArray(16384), 16384)
                while (!newSocket.isClosed) {
                    try {
                        newSocket.receive(receivedPacket)
                        log.debug("Received packet of size {} from {}", receivedPacket.length, receivedPacket.address.toString())
                        sizes.update(receivedPacket.length)
                        states[currentState]?.invoke(
                                ByteBuffer.wrap(receivedPacket.data, receivedPacket.offset, receivedPacket.length),
                                receivedPacket.socketAddress as InetSocketAddress
                        )
                    } catch (e: SocketTimeoutException) {
                        if (currentState == State.WAITING_FOR_PEER) {
                            callbacks.discoveryTimeout()
                            close()
                        } else if (e.message == "Socket closed") {
                            // do nothing, expected
                        }
                    } catch (e: Exception) {
                        log.error("Failure in receive loop", e)
                    }
                }
                log.info("Socket closed")
            })
        }
    }

    fun join(sessionId: String, fwdPort: Optional<Int> = Optional.empty()) {
        startIfRequired()

        callbacks.discovering()
        val joinCommand = fwdPort.map { "JOIN|" + sessionId + it }.orElse("JOIN|" + sessionId)

        currentState = State.WAITING_FOR_PEER
        currentSessionId = Optional.of(sessionId)
        socket.ifPresent { it.send(PacketUtils.buildPacket(joinCommand, discoveryServiceAddress)) }
    }

    fun send(data: RawData) {
        when {
            !peerConnection.isPresent ->
                log.error("Unable to send data, no peer connection")

            socket.map { it.isClosed }.orElse(false) ->
                log.error("Unable to send data, socket closed")

            else -> {
                val compressed = PacketUtils.compress(data.wrap.pack(serialiser))
                log.debug("Sending packet of size {} to {}", compressed.remaining(), peerConnection.get().toString())
                socket.get().send(DatagramPacket(compressed.array(), compressed.position(), compressed.remaining(), peerConnection.get()))
            }
        }
    }

    fun statistics(): Statistics = Statistics(
            latencies.snapshot.get99thPercentile().toInt(),
            outOfOrder,
            sizes.snapshot.get99thPercentile().toInt(),
            meter.oneMinuteRate
    )

    fun close() {
        if (currentState == State.WAITING_FOR_PEER && currentSessionId.isPresent) {
            socket.ifPresent { it.send(PacketUtils.buildPacket("CANCEL" + currentSessionId.get(), discoveryServiceAddress)) }
            reset()
            callbacks.discoveryCancelled()
        } else {
            reset()
            callbacks.closed()
        }
    }

    private fun reset() {
        if (socket.isPresent) {
            socket.get().close()
        }
        currentSessionId = Optional.empty()
        currentState = State.INITIAL
        peerConnection = Optional.empty()
    }
}

data class Statistics(val latency: Int, val outOfOrder: Int, val size: Int, val oneMinuteRate: Double)
